### DISCUSSION TITLE
sqlalchemy uses wrong syntax for ROLLUP on mariadb

### DISCUSSION THREAD

**Discussion by martin-greentrax** [sqlalchemy/sqlalchemy#8502] [Usage Questions] [answered]
Hello everybody,

I want to do a select-query with group_by + rollup. It works fine when I try my own raw sql in the db, but to me it seems sqlalchemy creates the wrong syntax (picks wrong dialect?) and it fails. Maybe I'm missing something in my config.

I'm using:
 sqlalchemy 1.4.41 on python 3.9
 and mariadb  Ver 15.1 Distrib 10.6.9-MariaDB

My database string looks like:

    mariadb+mariadbconnector://myuser:***@localhost/mydb

My python code:
  
    db.query(...).join(...).group_by(func.rollup(ClientDB.name, BookingDB.status))

It is created as:

    SELECT ... GROUP BY ROLLUP(client.name, booking.status)

which fails due to wrong syntax.

I would expect: `GROUP BY client.name, booking.status WITH ROLLUP`  
as per the mariadb documentation: https://mariadb.com/kb/en/select-with-rollup/

**Comment by CaselIT** [comment, marked as answer]
Hi,

I don't think that sintax is supported using functions, since rollup does not seem a function.

@zzzeek I know sqlalchemy has with hints, but I'm not sure we can render them in the groupby. If not then this should be filed as an use case issue

**Reply by zzzeek** [reply]
this would need to be a new non-func "rollup" construct and/or unary operator of some kind

i dont totally understand the mariadb syntax though as this seems to be a modifier for the GROUP BY overall, not for specific expressions, so we might want to just have `group_by(..., with_rollup=True)` if this is only a group_by thing (need to understand the syntax fully on all backends)


for now this can be done with a custom unary operator

```py
from sqlalchemy import Column
from sqlalchemy import Integer
from sqlalchemy import select
from sqlalchemy import UnaryExpression
from sqlalchemy.orm import declarative_base
from sqlalchemy.sql import operators
from sqlalchemy.sql.elements import UnaryExpression
Base = declarative_base()


class A(Base):
    __tablename__ = 'a'

    id = Column(Integer, primary_key=True)
    x = Column(Integer)
    y = Column(Integer)

def rollup(elem):
    return UnaryExpression(elem, modifier=operators.custom_op("WITH ROLLUP"))

stmt = select(A).group_by(A.x, rollup(A.y))

print(stmt)


```

**Reply by zzzeek** [reply]
it's an extension of the GROUP BY clause: https://www.sqltutorial.org/sql-rollup/ and it's only once

let's add `group_by(.., rollup=True)` as the use case

**Reply by martin-greentrax** [reply]
Thank you very much @zzzeek , I will try this out. Should I turn this discussion into an issue?

**Reply by zzzeek** [reply]
I guess a new feature req so we can track it ,. sure

**Reply by CaselIT** [reply]
I've taken the liberty of opening https://github.com/sqlalchemy/sqlalchemy/issues/8503
